Find a functionally equivalent code:com.android.internal.policy.impl.PhoneWindowManager.checkAddPermission:COMMENT
Method Modifier: public      internal    
Comment:/**
 * {@inheritDoc}
 */

@@ -7,18 +7,22 @@
     String permission = null;
     switch(type) {
         case TYPE_TOAST:
+            // XXX right now the app process has complete control over
+            // this...  should introduce a token to let the system
             // monitor/control what they are doing.
+            outAppOp[0] = AppOpsManager.OP_TOAST_WINDOW;
             break;
         case TYPE_DREAM:
         case TYPE_INPUT_METHOD:
         case TYPE_WALLPAPER:
         case TYPE_PRIVATE_PRESENTATION:
+        case TYPE_VOICE_INTERACTION:
             // The window manager will check these.
             break;
         case TYPE_PHONE:

